### ISSUE TITLE
There needs to be a way to toggle the default sort order when viewing a docket

### ISSUE THREAD

**Issue by troglodite2** [freelawproject/courtlistener#2628]
As an example of what I'm looking at:

![image](https://user-images.githubusercontent.com/82225842/230456710-13c008a2-832e-439d-a39b-0a5404260b35.png)

The position of the button isn't correct nor does it have any text.  I want to position it to the right of the sort buttons.

![image](https://user-images.githubusercontent.com/82225842/230456890-b4314d40-4e9f-4cdd-8165-6e40299e60e8.png)

With the tooltip over the button

![image](https://user-images.githubusercontent.com/82225842/230458221-3a52b2e9-221f-4b10-a520-bf8d1fa05e4b.png)

And after the button is toggled.

**Comment by mlissner** [comment]
Thanks for these screenshots. They helped me understand your concept and I took some time to talk about it with a few people to try to come up with better approach here that eliminates the need for any interface elements. 

A few notes:
 - Clicking the table headers should work too — This is a good point, but not one that I consider in scope for now.
 - Nobody came up with an interface that was particularly elegant. We discussed toggles, buttons like you have, a dropdown of some kind, a toast that pops up after you switch the ordering. All of it was either overly complex or added too much interface junk. 
 - Some users want all dockets sorted a certain way.
 - Some users want certain dockets always sorted one way and others another way.

Taken together, I think what I want to do is:
 - Keep the setting in the profile.
 - Keep track of dockets that the user has explicitly ordered a certain way. If my default is asc and I click to switch a docket to desc, then that docket should load that way, until I click the asc button again.
 - This means all logic shifts to the backend.

I think the architecture would probably require a two-column table where we store user ordering preferences, and then the ordering of the page would have a three-layer ordering decision:

1. If there's a order_by GET param, order it that way.
2. If the user has overridden their default, order it that way.
3. Otherwise, order by the user's default sort preference.

What do you think? Does this help give you a direction, and is the scope still within your willingness to build this feature?

-----

One other note: I noticed that the text in the profile is currently:

 - [ ] Sort dockets in descending order

I think it'd be more intuitive if it said:

 - [ ] Sort dockets in reverse-chronological order

Thank you!

**Comment by troglodite2** [comment]
Ok.  I think what we need is a many-to-many table.  For the time being let's call it "order_defaults".

```
LIST_CHOICES = [
  ('DOCKET','Docket'),
  ('OPINION','Opinion'),
  ('SEARCH','Search'),
]
DIR_CHOICES = [
  ('UP','Up'),
  ('DN','Down'),
]
user = models.ForeignKey(UserProfile)
list_name = models.CharField(
  max=10,
  choices=LIST_CHOICES,
  default='DOCKET',
)
order_by = models.CharField(
  max=2,
  choices=DIR_CHOICES,
  default='UP',
)
```

Please don't syntax check the above, it is pseudo code off the top of my head and I've not done any verification.

In the profile/account we use a field group and present one checkbox for each item in LIST_CHOICES.  When a user submits a filter with a `&order_by=` then we set or insert the entry into the order_default table.

This removes the need for any new user interface items.

Yes, this is within my skill set.  It is actually easier than all the work with htmx and CSS.

**Comment by mlissner** [comment]
I'm not sure we need a m2m, and I'm not sure why you're including opinions/search. I think the table would just be:

```
user = models.ForeignKey(User) # Not user profile
docket = models.ForeignKey(Docket)
order_preference = models.SmallIntegerField(choices=DIR_CHOICES)
```

A few notes:
 - This can FK to the User model, that tends to be faster/easier.
 - This can FK to the Docket model directly, without need for a m2m, I think?
 - I usually prefer SmallIntegerFields for choice fields. I think it's just a style thing for me?
 - I wouldn't want a default value on the `order_preference` field. We'll always set it explicitly instead.

What do you think?

**Comment by troglodite2** [comment]
1) We can link to the User instead of the UserPreference.  my thought was that we want the one with the fewest look ups.
2) There is no need for a FK to Docket.  We need a choice field there because I expect this to extend to all the other lists that are created.  In the view for that list it will be asking for `request.user.order_defaults.get(list_name='DOCKET')`
3) SmallIntegerField is my preferred but I know that many people prefer text. 
4) I put defaults in because it makes migrations safer.  Only reason.

**Comment by mlissner** [comment]
> I expect this to extend to all the other lists that are created. 

Gotcha. We'll do a migration for that when/if we ever build that, so let's count that out for now.

> my thought was that we want the one with the fewest look ups.

That's my goal too, which is why I usually join to the User, b/c then you can do `request.user.thing.filter()` as opposed to `request.user.profile.thing.filter()`. Feels like better practice to put this kind of thing off the root object too. But if you see it being more efficient on the Profile, go for it, and I'll be happy to be wrong. 

> There is no need for a FK to Docket. 

Hm, the idea is that each docket has its own ordering preference that can override the user default. So if I have DESC as my preference, then load a docket, it'll show up as DESC (of course). If I then click the ASC button, that'll store my preference *for that docket* to be ASC by default in the future.

**Comment by troglodite2** [comment]
Ok.  Not what I was seeing as the goal...

I am not adverse to building out the functionality you are describing but I'm not sure of the use case.  Do people really want to see the docket(1) the way they left it and docket(2) in a different order?

Consider three dockets.  You look at docket(1), click and make the ordering descending.  You know go to docket(2).  Do you expect to see docket(2) per your default setting or do you expect to see it the same order as the docket(1), your last viewed docket, or do you expect it to be in the order that you last viewed docket(2)?

I submit that we have two different issues that we are addressing.  
1. How do people want the different lists to be ordered?
2. What do people want to see when they change list order?

I suggest that we handle it as such:

Each user has a default for each list type, as I described above. 
When a user changes the sort order of that list type we retain that on a session variable for a limited time.  Call it 8 hours.
When the user goes to the next list of the same type it is ordered as it was on the last view if within that 8 hours, else it goes back to default.

My only concern is that time out.  I'm interested in feedback as to what people think the best option is.  To just maintain the last order as long as the user stays logged in or to expire it after a bit.

The con is that it makes the defaults almost unimportant.  The pro is that there isn't any database additions past the defaults.

At this point I'm going to recommend that we go with a default per list.  

Once we add accordions I believe that there will be fewer times that the user needs to change their default ordering.

**Comment by mlissner** [comment]
Well, I think different cases are reviewed differently. Some cases will be read top to bottom, others bottom to top, but you're probably right that I'm making it too complicated. 

I could see just saving this to the user's session and doing it the same way as they did last time, unless they're logged out. Their session would reset every two weeks and that'd probably be fine. If we do this, I think I'd suggest removing the feature you added to the profile, since it wouldn't be doing much.

What do you think?

**Comment by troglodite2** [comment]
Having discussed this here I'm going to propose that we put this issue on hold for the time being.  I'm going to create a PR for the accordions and I think that once that is in place and we have a chance to see how that works for people we can readdress this issue.

I believe that once we have played with the accordions a little bit that we'll be in a better position to know what the better path forward might be.
